util/lockwaiter/lockwaiter.go

```diff
@@ -115,56 +114,69 @@ func (lw *Manager) NewWaiter(startTS, lockTS, keyHash uint64, timeout time.Durat
 
 // WakeUp wakes up waiters that waiting on the transaction.
 func (lw *Manager) WakeUp(txn, commitTS uint64, keyHashes []uint64) {
+	var (
+		waiters    []*Waiter
+		remainSize int
+	)
 	lw.mu.Lock()
 	q := lw.waitingQueues[txn]
-	lw.mu.Unlock()
 	if q != nil {
 		sort.Slice(keyHashes, func(i, j int) bool {
 			return keyHashes[i] < keyHashes[j]
 		})
-		waiters, remainSize := q.getReadyWaiters(keyHashes)
+		waiters, remainSize = q.getReadyWaiters(keyHashes)
 		if remainSize == 0 {
-			lw.mu.Lock()
 			delete(lw.waitingQueues, txn)
-			lw.mu.Unlock()
 		}
-		sort.Slice(waiters, func(i, j int) bool {
-			return waiters[i].startTS < waiters[j].startTS
-		})
+	}
+	lw.mu.Unlock()
+
+	// wake up waiters
+	if len(waiters) > 0 {
 		for i, w := range waiters {
-			w.ch <- Result{Position: Position(i), CommitTS: commitTS}
+			w.ch <- WaitResult{Position: Position(i), CommitTS: commitTS}
 		}
-		log.Info("wakeup", len(waiters), "txns blocked by", txn, keyHashes)
+		log.Info("wakeup", len(waiters), "txns blocked by txn", txn, " keyHashes=", keyHashes)
 	}
 }
 
 // CleanUp removes a waiter from waitingQueues when wait timeout.
 func (lw *Manager) CleanUp(w *Waiter) {
 	lw.mu.Lock()
 	q := lw.waitingQueues[w.LockTS]
-	lw.mu.Unlock()
 	if q != nil {
 		q.removeWaiter(w)
```

Comment:
Need to check q is empty and remove from the map

Reply:
Added